include/maths/CTools.h

```diff
@@ -668,6 +668,17 @@ class MATHS_EXPORT CTools : private core::CNonInstantiatable {
     //! Compute \f$x^2\f$.
     static double pow2(double x) { return x * x; }
 
+    //! Compute a value from \p x which will be stable across platforms.
+    static double stable(double x) {
+        return STABLE_EPS * std::floor(x / STABLE_EPS + 0.5);
```

Comment:
Could this cause problems if `x` is very large?  This is dividing by a number of the order of 1e-14 right?  So if `x` was 1e300, would anything go wrong?

Just to prove there are no nasty edge cases it would be good to add a unit test that exercises this function for these arguments:

* -1e305
* -1.0
* -1e-305
* 0.0
* 1e-305
* 1.0
* 1e305

Reply:
This is a good point. It won't be near overflowing given how it's used, but, in fact, we have some tools for dealing with this correctly already I should have used.

Reply:
Ok I just completely reworked this to drop the last bit in a way which doesn't rely on any standard library functions: using `frexp` and `ldexp` defeats the purpose. This implementation clearly doesn't overflow, so instead I added some tests for bitwise comparison of `stableLog` and `stableExp` for some random inputs. These would fail if they're different on any of our target platforms. See [7803c5c](https://github.com/elastic/ml-cpp/pull/1127/commits/7803c5c31fabb352f764b76b99885ad6f14382a2).